### ISSUE TITLE
[Question] Vogen in "dynamic" EntityFramework Core situations

### ISSUE THREAD

**Issue by mkalinski93** [SteveDunn/Vogen#699]
### Describe the feature

Hello,

I´m working on a project, which heavily uses Vogen, that needs to filter values submitted by the FrontEnd, on the BackEnd with IQueryable.

I would say that 95% of the cases are doing great and I´m enjoying the work with Vogen so far.
Now I´ve encountered one specific type of problem that is drastically putting me into the corner.
I need to filter values with a "Contains" or "IN" operation. For example, I have a List of Guid/string supplied from the frontEnd that needs to filter the property 'Id: UserId' (that´s my ValueObject here). I will receive the property name "id" and a collection of Guid/String values.
The real problem starts here: Whenever I "dynamically" want to filter the database, I have limited choices:

1. Use a custom library that parses strings into Expression - like Gridify, Dynamic.LINQ and so on
2. Implement my own library - I´m not scared of
3. Implement some sort of "EF.Property" butchery - not a fan of

But all three choices brought me back to a single point of failure - Type safety.
If the QueryFilter object doesn´t explicitely contains the type of UserId, either as an object or as a generic collection, I´m done. I would need to explicitely specify the type, otherwise EfCore cannot work with it.

Whenever I try to provide some sort of "dynamic filter", the ValueConverter of EfCore will be in charge, converting the values back and forth (Primitive <-> ValueObject) and I cannot use primitive types from the request as it expects the real ValueObject with correct type. The result will always be empty. And it makes sense. Now, after a few days of pain, I finally want to reliefe myself and write down this question. Is there anybody here, that has found a solution to it?

**Comment by SteveDunn** [comment]
Hi, thanks for starting this discussion. It sounds like a common enough scenario, so I'm sure we can get something working.

Would you be able to provide a minimal repro that I could compile and run to see what needs to be done.

**Comment by ErgEnn** [comment]
Hi!

While I am not the OP, I think I recreated a minimal repro that explains the problem.
https://gist.github.com/ErgEnn/ae8bda788bdc9c4fb5807774510d5d35

**Comment by SteveDunn** [comment]
Thank you @ErgEnn - I'll try to get around to this soon.

**Comment by JoschiZ** [comment]
This is an old issue, but I ran into this problem before and how you can work around it / solve it is not that intuitive.
So maybe this helps someone or it can be included in the docs.

I'm also not sure if Vogen can do anything about this.

TLDR: you need to cast the value object to it's underlying type:

``` c#
// if explicit casts are enabled
context.Items.Single(item => ((string)item.ValueObject).Contains("e"));

// otherwise just do it doesn't matter that this would throw at runtime, because it is not being run, just translated
context.Items.Single(item => ((string)(object)item.ValueObject).Contains("e"));
```

#### The Problem
The underlying problem is that EFCore is trying to parse an expression tree.

``` c#
context.Items.Single(item => item.StringValue.Contains("e"))
```

works fine, because EFCore can translate the known method `string::Contains` into the corresponding SQL syntax.
Something like 

``` sql
SELECT * FROM [Entities]
WHERE [StringValue] LIKE '%e%'
```

But now you try to do the same with an Value object (of any kind).

``` c#
context.Items.Single(item => item.StringValueObject.Value.Contains("e"))
```

we need to use `.Value.Contains` because the value object has no `Contains` implementation.
And even if it had one, it would not be recognized by EFCore.

And this `.Value` is the culprit of why EFCore cannot translate that expression.
Which column would `.Value` correspond to in the `Entities` table? It's completely unknown to EFCore.

But because EFCore just parses the expression tree and is not actually running our code, we can do some shenanigans.

Casting the value object to it's underlying type does the trick

``` c#
// if explicit casts are enabled
context.Items.Single(item => ((string)item.ValueObject).Contains("e"));

// otherwise just do it doesn't matter that this would throw at runtime, because it is not being run, just translated
context.Items.Single(item => ((string)(object)item.ValueObject).Contains("e"));
```

**Comment by SteveDunn** [comment]
Great - thank you for input, I'll take a look at the PR - looks very ueful!
